src/value.rs

```diff
@@ -5,69 +5,231 @@
 pub struct Value {
-    inner: ValueInner,
+    /// The underlying data, in "padded" bit-encoded form, with the leftmost
+    /// bits encoded in the most-significant bits. So e.g. right(unit) is 0x80.
+    ///
+    /// We use the padded form, even though it is space-inefficient (sometimes
+    /// significantly so), for a couple reasons: first, this is the representation
+    /// used by the bit machine, so we want encoding/decoding to be as fast as
+    /// possible. Secondly, when destructuring a value (which we need to do during
+    /// scribing and pruning, at least) it is quite difficult to destructure a
+    /// compact-encoded product, since we would need to determine the compact
+    /// bitlengths of the children, and these quantities are not readily available.
+    inner: Arc<[u8]>,
+    /// An offset, in bits, at which the actual data starts. This is useful
+    /// because it allows constructing sub-values of a value without needing
+    /// to construct a new `inner` with all the bits offset.
+    bit_offset: usize,
+    /// The Simplicity type of the value.
     ty: Arc<Final>,
 }
 
-/// The inner structure of a Simplicity value.
-#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-enum ValueInner {
-    /// The unit value.
-    ///
-    /// The unit value is the only value of the unit type `1`.
-    /// It must be wrapped in left and right values to encode information.
-    Unit,
-    /// A left value.
-    ///
-    /// A left value wraps a value of type `A` and its type is the sum `A + B` for some `B`.
-    /// A `false` bit encodes that a left value is wrapped.
-    Left(Arc<Value>),
-    /// A right value.
-    ///
-    /// A right value wraps a value of type `B` and its type is the sum `A + B` for some `A`.
-    /// A `true` bit encodes that a right value is wrapped.
-    Right(Arc<Value>),
-    /// A product value.
-    ///
-    /// A product value wraps a left value of type `A` and a right value of type `B`,
-    /// and its type is the product `A × B`.
-    /// A product value combines the information of its inner values.
-    Product(Arc<Value>, Arc<Value>),
+// Because two equal values may have different bit offsets, we must manually
+// implement the comparison traits. We do so by first comparing types, which
+// is constant overhead (this just compares TMRs). If those match, we know
+// the lengths and structures match, so we then compare the underlying byte
+// iterators.
+impl PartialEq for Value {
+    fn eq(&self, other: &Self) -> bool {
+        self.ty == other.ty && self.raw_byte_iter().eq(other.raw_byte_iter())
+    }
+}
+impl Eq for Value {}
+
+impl PartialOrd for Value {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Ord for Value {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        self.ty
+            .cmp(&other.ty)
+            .then_with(|| self.raw_byte_iter().cmp(other.raw_byte_iter()))
+    }
+}
```

Comment:
cbbe9b5a3e7982c5b4c85b28cc37996fbcb0c7b2: In a follow-up PR, we should remove the `Ord` implementation. It makes no semantic sense to compare values of different types. Rather, `PartialOrd` could return `None` if `self.ty != other.ty`.

Reply:
Yes -- I tried this, but we then lose `Ord` on `Word` (which maaybe makes sense, you can compare them numerically though you have to decide the relationship between 0u4 and 0u16 etc), and on `Node` and on `Policy`... so it's far-reaching.

I think the right approach would be to change `Policy` to implement `Ord` manually in alphabetical order (we do this in rust-miniscript now; it was a PITA to implement) and then we can drop `Ord` on the smaller types.

In any case, out of scope for this PR.